[SPARK-27936][K8S] Support python deps

skonto opened the pull request:
### What changes were proposed in this pull request?
Supports python client deps from the launcher fs.
### Why are the changes needed?
This is a feature that was added for java deps. This PR adds support fo rpythona s well.

### Does this PR introduce any user-facing change?
yes

### How was this patch tested?
Manually running different scenarios and via examining the driver & executors logs. Also there is an integration test added.
I verified that the python resources are added to the spark file server and they are named properly so they dont fail the executors. Note here that as previously the following will not work:
primary resource `A.py`: uses a closure defined in submited pyfile `B.py`, context.py only adds to the pythonpath files with certain extension eg. zip, egg, jar.

skonto committed:
support python deps

skonto commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -125,16 +125,16 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
+      .create(cephService))
 
-    kubernetesTestComponents
+    Eventually.eventually(TIMEOUT, INTERVAL) (kubernetesTestComponents
```
instead of watching for deletion make sure we are able to create the resources if some previous is still being deleted.

holdenk replied:
Maybe make that as a comment since most people won't go back to the PR to understand what is going.

skonto replied:
I added a comment above of each statement to clarify this.

SparkQA commented:
**[Test build #111078 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111078/testReport)** for PR 25870 at commit [`40fe23c`](https://github.com/apache/spark/commit/40fe23c540a8db7dcea335457dc8fe5d3328f529).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/16163/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/16163/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16163/
Test PASSed.

SparkQA commented:
**[Test build #111078 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111078/testReport)** for PR 25870 at commit [`40fe23c`](https://github.com/apache/spark/commit/40fe23c540a8db7dcea335457dc8fe5d3328f529).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111078/
Test PASSed.

skonto commented:
@erikerlandson @holdenk pls review. I know there are some other changes related to the tests, I can always rebase if they get merged first.

skonto commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -183,6 +185,50 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
+        .set("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+        .set("spark.jars.packages", "com.amazonaws:aws-java-sdk:" +
+          "1.7.4,org.apache.hadoop:hadoop-aws:2.7.6")
+        .set("spark.driver.extraJavaOptions", "-Divy.cache.dir=/tmp -Divy.home=/tmp")
```
I can refactor this part to have the properties set once as they are shared with the existing test. In general I think we should separate the Suites in the future to allow better setup for before and after conditions.

holdenk commented:
cc @ifilonenko who I think did some related work?

skonto commented:
@erikerlandson if you have some free time pls have a look :)

erikerlandson commented:
@skonto this LGTM

holdenk commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -259,12 +259,19 @@ private[spark] object KubernetesUtils extends Logging {
+      isJava: Boolean = true): String = {
     if (isLocalAndResolvable(resource)) {
-      SparkLauncher.NO_RESOURCE
+      if (isJava) {
```
I think I'm missing some context, and others may be as well. Could we add a comment about why we need to upload the non-Java main resource?

skonto replied:
I upload it because it resided at the client machine

holdenk replied:
So I guess the part I don't understand is why would it not need to be uploaded if it was Java? Presumably we upload the java somewhere else, why isn't the Python being uploaded in the same place?

erikerlandson replied:
@holdenk does the java loading explaination work for you:
https://github.com/apache/spark/pull/25870#issuecomment-541001902

To confirm - the timing of the java loading is to ensure java code can execute when its needed?

holdenk replied:
I think so, I was hoping we'd unify the code paths a bit more but that make sense.

holdenk reviewed (commented):
Follow up question re: uploading.

skonto commented:
@holdenk this is because spark-submit add the resource in sparks.jars:
```
19/10/11 13:01:30 WARN Utils: Your hostname, universe resolves to a loopback address: 127.0.1.1; using 192.168.2.4 instead (on interface wlp2s0)
19/10/11 13:01:30 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Parsed arguments:
  master                  k8s://https://10.96.0.1:443
  deployMode              cluster
  executorMemory          1G
  executorCores           null
  totalExecutorCores      null
  propertiesFile          null
  driverMemory            1G
  driverCores             null
  driverExtraClassPath    null
  driverExtraLibraryPath  null
  driverExtraJavaOptions  null
  supervise               false
  queue                   null
  numExecutors            2
  files                   null
  pyFiles                 null
  archives                null
  mainClass               org.apache.spark.examples.SparkPi
  primaryResource         local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar
  name                    spark-pi
  childArgs               [100]
  jars                    null
  packages                null
  packagesExclusions      null
  repositories            null
  verbose                 true

Spark properties used, including those specified through
 --conf and those from the properties file null:
  (spark.kubernetes.driver.pod.name,spark-pi-driver)
  (spark.executor.instances,2)
  (spark.driver.memory,1G)
  (spark.executor.memory,1G)
  (spark.kubernetes.authenticate.driver.serviceAccountName,spark-sa)
  (spark.kubernetes.namespace,spark)
  (spark.kubernetes.container.image,lightbend/spark:2.1.2-OpenShift-2.4.4-rh-2.12)
  (spark.kubernetes.container.image.pullPolicy,Always)

    
19/10/11 13:01:31 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Main class:
org.apache.spark.deploy.k8s.submit.KubernetesClientApplication
Arguments:
--primary-java-resource
local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar
--main-class
org.apache.spark.examples.SparkPi
--arg
100
Spark config:
(spark.kubernetes.namespace,spark)
(spark.jars,local:///opt/spark/examples/jars/spark-examples_2.12-2.1.2-2.4.4-lightbend.jar)
(spark.app.name,spark-pi)
(spark.driver.memory,1G)
(spark.executor.instances,2)
(spark.submit.pyFiles,)
(spark.kubernetes.container.image.pullPolicy,Always)
(spark.kubernetes.container.image,lightbend/spark:2.1.2-OpenShift-2.4.4-rh-2.12)
(spark.submit.deployMode,cluster)
(spark.master,k8s://https://10.96.0.1:443)
(spark.kubernetes.authenticate.driver.serviceAccountName,spark-sa)
(spark.executor.memory,1G)
(spark.kubernetes.driver.pod.name,spark-pi-driver)
Classpath elements:
```
So since the PR [here](https://github.com/apache/spark/pull/23546) manages that property we dont need to do double work. For python its different.

SparkQA commented:
**[Test build #112169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112169/testReport)** for PR 25870 at commit [`3245e4e`](https://github.com/apache/spark/commit/3245e4ec858822eb265f02da99f011218079a6a9).

SparkQA commented:
**[Test build #112171 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112171/testReport)** for PR 25870 at commit [`4ed8524`](https://github.com/apache/spark/commit/4ed85247fe5483a9975adddf6c79666aadc47a3d).

erikerlandson commented:
> @erikerlandson I added one test which uses a zip file as we discussed previously:

Thanks @skonto!

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17161/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17161/

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17161/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17163/

SparkQA commented:
**[Test build #112171 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112171/testReport)** for PR 25870 at commit [`4ed8524`](https://github.com/apache/spark/commit/4ed85247fe5483a9975adddf6c79666aadc47a3d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112171/
Test PASSed.

SparkQA commented:
**[Test build #112169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112169/testReport)** for PR 25870 at commit [`3245e4e`](https://github.com/apache/spark/commit/3245e4ec858822eb265f02da99f011218079a6a9).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112169/
Test PASSed.

holdenk commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
+      val (accessKey, secretKey) = getCephCredentials()
+      sparkAppConf
+        .set("spark.kubernetes.container.image", pyImage)
+        .set("spark.kubernetes.pyspark.pythonVersion", "2")
```
Lets not use Python 2 since this is targeted for Spark 3 I believe.

skonto replied:
I got used to 2 for years will miss it :) Sure will change.

holdenk reviewed (commented):
Minor comment about Python 2, but really excited to see this.

SparkQA commented:
**[Test build #112275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112275/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17258/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17258/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17258/
Test FAILed.

SparkQA commented:
**[Test build #112275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112275/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

skonto commented:
@erikerlandson can I get a merge? Gentle ping :)

skonto commented:
@holdenk gentl ping.

skonto commented:
@dongjoon-hyun could you merge this one?

dongjoon-hyun commented:
Thank you for pinging me, @skonto . I'll take a look now.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #112745 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112745/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).

SparkQA commented:
**[Test build #112745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112745/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112745/
Test FAILed.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala:
```diff
@@ -259,12 +259,19 @@ private[spark] object KubernetesUtils extends Logging {
+  def renameMainAppResource(
+      resource: String,
+      conf: Option[SparkConf] = None,
+      isJava: Boolean = true): String = {
```
Can we use `language-neural` parameter name, @skonto ? Then, this function can be reused for `SparkR` later.

dongjoon-hyun replied:
Also, let's not have the default value. This function is used only twice. And you changed both places in this PR. There is no reason to introduce a new default value.

skonto replied:
@dongjoon-hyun ok sure.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17690/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/17690/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17690/
Test FAILed.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverCommandFeatureStep.scala:
```diff
@@ -73,7 +76,9 @@ private[spark] class DriverCommandFeatureStep(conf: KubernetesDriverConf)
 
-    val pythonContainer = baseDriverContainer(pod, res)
+    // re-write primary resource to be the remote one and upload the related file
+    val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), false)
```
We recommend the following stye.
```scala
- val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), false)
+ val newResName = KubernetesUtils.renameMainAppResource(res, Some(conf.sparkConf), isJava = false)
```
The above is an example. When we rename `isJava`, we should change here together.

skonto replied:
ok

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala:
```diff
@@ -170,4 +170,3 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
     additionalProps.toMap
   }
 }
-
```
I understand the intention, but let's not touch irrelevant part.

skonto replied:
ok

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -125,16 +125,18 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
 
-    kubernetesTestComponents
+    // try until the service from a previous test is deleted
+    Eventually.eventually(TIMEOUT, INTERVAL) (kubernetesTestComponents
```
Is this relevant to this PR? If this aims to the flakiness of the existing test suite, we should make another PR before this one.

skonto replied:
Nope this is to address waiting for the elimination of the running service from a previous test. It is an easy retry until the previous service is removed and the new can succeed.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -125,16 +125,18 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
+      .create(cephService))
 
-    kubernetesTestComponents
+    // try until the stateful set of a previous test is deleted
```
ditto.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -143,24 +145,26 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
       .kubernetesClient
       .services()
       .withName(svcName)
+      .withGracePeriod(0)
```
ditto for the above two `.withGracePeriod(0)`s.

skonto replied:
We want to remove resources asap so grace period forces it, we dont want that?

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
     }
   }
 
+  test("Launcher python client dependencies using py", k8sTestTag, MinikubeTag) {
```
`using py`?

skonto replied:
yes correct.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:
```diff
@@ -183,6 +187,62 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
+        .set("spark.jars.packages", "com.amazonaws:aws-java-sdk:" +
+          "1.7.4,org.apache.hadoop:hadoop-aws:2.7.6")
+        .set("spark.driver.extraJavaOptions", "-Divy.cache.dir=/tmp -Divy.home=/tmp")
+      createS3Bucket(accessKey, secretKey, cephUrlStr)
```
+1 for refactor from line 204 ~ 224. We had better avoid this kind duplication.
As you suggested, please make this part as a utility function for the existing test and this new test.

skonto replied:
ok.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
     }
   }
+
+  def getTestFileAbsolutePath(fileName: String, sparkHomeDir: Path): String = {
```
@skonto . In this PR, you copied `getExamplesJarAbsolutePath` to `getTestFileAbsolutePath` and removed all instances using `getExamplesJarAbsolutePath`. Do we think we need to maintain the unused function, `getExamplesJarAbsolutePath`?

In this case, I'd like to recommend you to generalize the existing function by renaming.

skonto replied:
Hmmm let me check.

skonto replied:
You are right we dont need it anymore.

skonto replied:
You are right we dont need it anymore, I will remove it.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
+    }
+  }
 }
+
```
In one file, you remove the empty line at the end.
- https://github.com/apache/spark/pull/25870/files#diff-7effa575514dd61752ece0b40fc2ad23L173
And here, you are adding a new empty line. 😄 .

Let's remove this redundant empty line.

skonto replied:
@dongjoon-hyun did I ok :) Usually I add a \n at the end as an old practice: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file
https://thoughtbot.com/blog/no-newline-at-end-of-file

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:
```diff
@@ -131,4 +133,41 @@ object Utils extends Logging {
+      IOUtils.closeQuietly(fis)
+      IOUtils.closeQuietly(zipOut)
+    } catch {
+      case e: Exception => log.error(s"Failed to zip file: $inFile"); throw e
```
Do we need to have `log.error` additionally when we re-throw the same `e`? Since the test case will fail eventually, I think we can skip the whole `try ... catch` effort.

skonto replied:
I added this so I can print the message for the failure. I can remove and see how informative the exception is.

dongjoon-hyun reviewed (changes requested):
Hi, @skonto . I finished my first round review. I'll review again after updating.

skonto commented:
Ok I will fix this over the weekend, sorry for the delay.

skonto commented:
@dongjoon-hyun  could you re-open it pls? I will work on it.

dongjoon-hyun commented:
Sure, no problem. Thanks, @skonto .

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25908/
Test FAILed.

SparkQA commented:
**[Test build #121220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121220/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).

SparkQA commented:
**[Test build #121220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121220/testReport)** for PR 25870 at commit [`f8df227`](https://github.com/apache/spark/commit/f8df2273f426a92c6742b1520d7fc480ba75c975).
 * This patch **fails build dependency tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121220/
Test FAILed.

SparkQA commented:
**[Test build #122325 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122325/testReport)** for PR 25870 at commit [`1959956`](https://github.com/apache/spark/commit/19599562625ca331b619bebe54b47f54e2a9621f).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/26995/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/26995/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26995/
Test PASSed.

SparkQA commented:
**[Test build #122325 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122325/testReport)** for PR 25870 at commit [`1959956`](https://github.com/apache/spark/commit/19599562625ca331b619bebe54b47f54e2a9621f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122325/
Test PASSed.

SparkQA commented:
**[Test build #122415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122415/testReport)** for PR 25870 at commit [`7fca5e4`](https://github.com/apache/spark/commit/7fca5e433fe0bb49259d4bbf1ecb6aff626707f6).

SparkQA commented:
**[Test build #122415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122415/testReport)** for PR 25870 at commit [`7fca5e4`](https://github.com/apache/spark/commit/7fca5e433fe0bb49259d4bbf1ecb6aff626707f6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122415/
Test FAILed.

SparkQA commented:
**[Test build #122416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122416/testReport)** for PR 25870 at commit [`b0bc40f`](https://github.com/apache/spark/commit/b0bc40fd611898a6906629465ddb88d8f31af5b4).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27079/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27080/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27079/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27079/
Test PASSed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27080/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27080/
Test FAILed.

skonto commented:
@dongjoon-hyun ready for a review round. Refactored a few stuff. Thank you for your patience :)

dongjoon-hyun commented:
@skonto . I visited this PR two days ago when you refreshed and visited now again.
But, my comments are still not addressed. That's the reason I didn't review this again.

SparkQA commented:
**[Test build #122416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122416/testReport)** for PR 25870 at commit [`b0bc40f`](https://github.com/apache/spark/commit/b0bc40fd611898a6906629465ddb88d8f31af5b4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122416/
Test FAILed.

skonto commented:
@dongjoon-hyun sure will have a look.

SparkQA commented:
**[Test build #122444 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122444/testReport)** for PR 25870 at commit [`eb67283`](https://github.com/apache/spark/commit/eb67283c124518fb44943e7dfcff9033d3c0d6f6).

SparkQA commented:
**[Test build #122444 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122444/testReport)** for PR 25870 at commit [`eb67283`](https://github.com/apache/spark/commit/eb67283c124518fb44943e7dfcff9033d3c0d6f6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122444/
Test FAILed.

SparkQA commented:
**[Test build #122445 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122445/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27105/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27106/

skonto commented:
@dongjoon-hyun I think addressed your comments, I missed a few earlier, had to scroll up a bit. There is a discussion about the use of `Eventually` but I have explained why I used this way to make sure the resources are cleaned up. Same for the use of the grace period there, it is no harm afaik.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27105/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27105/
Test PASSed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27106/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27106/
Test PASSed.

SparkQA commented:
**[Test build #122445 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122445/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122445/
Test PASSed.

skonto commented:
@dongjoon-hyun gentle ping, pls let me know if I need to do anything more.

dongjoon-hyun commented:
Very sorry for the delay. Since last Friday, I was very busy for the other stuffs. (The Jenkins outage and SparkSummit slide due). Since this is for 3.1.0, I'll try to re-visit here before this weekend. Hopefully, sooner than that.

dongjoon-hyun commented:
Retest this please

SparkQA commented:
**[Test build #122599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122599/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
**[Test build #122599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122599/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122599/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27261/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27261/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27261/
Test FAILed.

skonto commented:
@dongjoon-hyun thanks! I checked the failures but dont seem to be relevant. I will try re-test.

skonto commented:
Retest this please

SparkQA commented:
**[Test build #122630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122630/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
**[Test build #122630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122630/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122630/
Test FAILed.

dongjoon-hyun commented:
The dependency issue is fixed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #122647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122647/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
**[Test build #122647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122647/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122647/
Test FAILed.

dongjoon-hyun commented:
The job was running on `amp-jenkins-worker-05` which seems to have a corrupted Maven local info. This is another issue.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-cli) on project spark-parent_2.12: ArtifactInstallerException: Failed to install metadata org.apache.spark:spark-parent_2.12/maven-metadata.xml: Could not parse metadata /home/jenkins/.m2/repository/org/apache/spark/spark-parent_2.12/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got > (position: END_TAG seen ...</metadata>\n>... @13:2) -> [Help 1]
```

dongjoon-hyun commented:
Retest this please

SparkQA commented:
**[Test build #122650 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122650/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
**[Test build #122650 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122650/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122650/
Test FAILed.

dongjoon-hyun commented:
It's still assigned to `amp-jenkins-worker-05`.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27309/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27312/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27309/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27309/
Test FAILed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27312/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27312/
Test FAILed.

dongjoon-hyun commented:
Retest this please

SparkQA commented:
**[Test build #122747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122747/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27406/

SparkQA commented:
**[Test build #122747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/122747/testReport)** for PR 25870 at commit [`a9f055b`](https://github.com/apache/spark/commit/a9f055b4642d487b724a628c4d6f682c71fd2d2c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/122747/
Test PASSed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/27406/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27406/
Test FAILed.

skonto commented:
Back to this.

holdenk commented:
I'm also back after my motorcycle crash last year, let me know when it's up to date and I'd be happy to take another look. Thanks for sticking with this.

erikerlandson commented:
I notice this is targeted for 3.1, is it worth trying to get on the 3.0.1 train?

holdenk commented:
I think the plan is to cut 3.0.1 pretty soon since there are some almost show stopper bugs in 3.0 so I’d leave it targeted to 3.1 and once it’s in we can discuss if backporting makes sense.

dongjoon-hyun commented:
+1 for @holdenk 's comment. Also, in general, we cannot backport the improvement JIRA.

dongjoon-hyun commented:
Gentle ping, @skonto . Apache Spark 3.1.0 Feature Freeze is scheduled on Early Nov 2020
- https://spark.apache.org/versioning-policy.html

skonto commented:
@dongjoon-hyun I will try update, sorry it has been too long, my bad.

dongjoon-hyun commented:
Gentle ping, @skonto .

skonto commented:
ack

Commit:
update dep it

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35889/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35889/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35889/
Test PASSed.

skonto commented:
@dongjoon-hyun I finally updated it pls review.

SparkQA commented:
**[Test build #131286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131286/testReport)** for PR 25870 at commit [`0145183`](https://github.com/apache/spark/commit/0145183265a1a27c38cd853c3bba53df0b8552e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131286/
Test PASSed.

dongjoon-hyun commented:
Thank you so much for updates, @skonto !

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:
```diff
@@ -114,28 +116,39 @@ object Utils extends Logging {
   }
+
+  def createZipFile(inFile: String, outFile: String): Unit = {
+      val fileToZip = new File(inFile)
```
indentation?

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @skonto and all.
I'll fix the indentation and merge this.

Merged to master for Apache Spark 3.1.0.

skonto commented:
Just saw your comment thanks!!!
